src/url.rs

```diff
@@ -29,35 +85,46 @@ impl Scanner for UrlScanner {
 }
 
 impl UrlScanner {
+    // For URL searching starting before the `://` separator, the `has_scheme` parameter should be
+    // true because the URL will have a scheme for sure. If seraching before the `.` separator, it
+    // should be `false` as we might search over the scheme definition for the scheme being optional.
     // See "scheme" in RFC 3986
-    fn find_start(&self, s: &str) -> (Option<usize>, Option<char>) {
+    fn find_start(&self, s: &str, mut has_scheme: bool) -> (Option<usize>, Option<char>) {
         let mut first = None;
-        let mut digit = None;
+        let mut special = None;
         let mut quote = None;
         for (i, c) in s.char_indices().rev() {
             match c {
                 'a'..='z' | 'A'..='Z' => first = Some(i),
-                '0'..='9' => digit = Some(i),
-                // scheme special
+                '0'..='9' => special = Some(i),
+                '/' if !has_scheme => special = Some(i),
+                ':' if !has_scheme => {
+                    has_scheme = true;
+                    special = Some(i)
+                }
                 '+' | '-' | '.' => {}
-                _ => {
+                '@' => return (None, None),
+                c if QUOTES.contains(&c) => {
                     // Check if there's a quote before the scheme,
                     // and stop once we encounter one of those quotes.
                     // https://github.com/robinst/linkify/issues/20
-                    if QUOTES.contains(&c) {
-                        quote = Some(c);
-                    }
-                    break;
+                    quote = Some(c);
+                }
+                c if !has_scheme && !matches!(c, '(' | ')' | '[' | ']' | '{' | '}' | ' ') => {
```

Comment:
Somewhat ugly, do you think this is fine?

Reply:
Honestly I think it's fine. If anything, we could move the `matches` part into a separate function (and add an inline attribute). The question is what we'd call the function. 😄 

```rust
#[inline]
/// Checks against characters that can't be the start of a URL
fn is_special_character(c: &char) -> bool {
    matches!(c, '(' | ')' | '[' | ']' | '{' | '}' | ' ')
}
```

Reply:
Right! Some sort of negated pattern would have been great here, but there isn't a nice way to express that other than what I've done now, as far as I know.